app/questionnaire/questionnaire_store_updater.py

```diff
@@ -18,20 +18,50 @@ def __init__(self, current_location, schema, questionnaire_store, current_questi
         self._list_store = self._questionnaire_store.list_store
         self._completed_items = self._questionnaire_store.completed_items
 
-    def save_answers(self, form, save_completed_blocks=True):
+    def save(self):
+        self._questionnaire_store.add_or_update()
```

Comment:
do we need dirty checking? so we only action the save if something has actually changed?

Reply:
Good spot. I've converted CompletedItems into CompletedStore and added is_dirty checking to it and the ListStore. The updater now checks the is_dirty status of all three stores before deciding to save.